quakemigrate/io/data.py

```diff
@@ -413,6 +398,92 @@ def __init__(self, starttime, endtime, sampling_rate, stations=None,
+        """
+        Check waveform availability against data quality criteria. There are a
+        number of hard-coded checks: for whether any data is present; for
+        whether the data is a flatline (all samples have the same value); and
+        for whether the data contains overlaps. There are a selection of
+        additional optional checks which can be specified according to the
+        onset function / user preference.
+
+        Parameters
+        ----------
+        st : `obspy.Stream` object
+            Stream containing the waveform data to check against the
+            availability criteria.
+        all_channels : bool, optional
+            Whether all supplied channels (distinguished by SEED id) need to
+            meet the availability criteria to mark the data as 'available'.
+        allow_gaps : bool, optional
+            Whether to allow gaps.
+        full_timespan : bool, optional
+            Whether to ensure the data covers the entire timespan requested;
+            note that this implicitly requires that there be no gaps.
+
+        Returns
+        -------
+        available : int
+            0 if data doesn't meet the availability requirements; 1 if it does.
+        availability : dict
+            Dict of {tr_id : available} for each unique SEED ID in the input
+            stream (available is again 0 or 1).
+
+        """
+
+        availability = {}
+
+        # Check if any channels in stream
+        if bool(st):
+            # Loop through channels with unique SEED id's
+            for tr_id in sorted([tr.id for tr in st]):
+                st_id = st.select(id=tr_id)
+                availability[tr_id] = 1
+
+                # First check data is here:
+                if not bool(st_id):
+                    availability[tr_id] = 0
+                    continue
+                # Check it's not flatlined
+                if any([tr.data.max() == tr.data.min() for tr in st_id]):
+                    availability[tr_id] = 0
+                    continue
+                # Check for overlaps
+                overlaps = st_id.get_gaps(max_gap=-0.000001)
+                if len(overlaps) != 0:
+                    availability[tr_id] = 0
+                    continue
+                # Check for gaps (if requested)
+                if not allow_gaps:
+                    gaps = st_id.get_gaps()  # Overlaps already dealt with
+                    if len(gaps) != 0:
+                        availability[tr_id] = 0
+                        continue
+                # Check data covers full timespan (if requested)
+                if full_timespan:
+                    if len(st_id) > 1:
+                        availability[tr_id] = 0
+                    else:
+                        tr = st_id[0]
+                        if tr.stats.starttime != self.starttime or \
+                            tr.stats.endtime != self.endtime:
+                            availability[tr_id] = 0
```

Comment:
```suggestion
                    if len(st_id) > 1:
                        availability[tr_id] = 0
                    elif st_id[0].stats.starttime != self.starttime or \
                        st_id[0].stats.endtime != self.endtime:
                            availability[tr_id] = 0
```

Reply:
Also, in what cases does the first test ever fail? Surely if there are multiple streams they must either be overlapping or gapped, hence already handled?

Reply:
Gaps only handled already if `allow_gaps=False`